controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,410 @@
+
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=openstackfloatingippools,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=openstackfloatingippools/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ipam.cluster.x-k8s.io,resources=ipaddressclaims;ipaddressclaims/status,verbs=get;list;watch;update;create;delete
+// +kubebuilder:rbac:groups=ipam.cluster.x-k8s.io,resources=ipaddresses;ipaddresses/status,verbs=get;list;watch;create;update;delete
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *OpenStackFloatingIPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+	pool := &infrav1.OpenStackFloatingIPPool{}
+	if err := r.Client.Get(ctx, req.NamespacedName, pool); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	scope, err := r.ScopeFactory.NewClientScopeFromFloatingIPPool(ctx, r.Client, pool, r.CaCertificates, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if pool.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Add finalizer if it does not exist
+		if controllerutil.AddFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+			return ctrl.Result{}, r.Client.Update(ctx, pool)
+		}
+	} else {
+		// Handle deletion
+		return ctrl.Result{}, r.reconcileDelete(ctx, scope, pool)
+	}
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	if err := r.reconcileFloatingIPNetwork(scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.setIPStatuses(ctx, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+		if !claim.ObjectMeta.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			ipAddress := &ipamv1.IPAddress{}
+			if err := r.Client.Get(ctx, client.ObjectKey{Name: claim.Name, Namespace: claim.Namespace}, ipAddress); err == nil {
+				// IPAddress already exists, another reconciler is working on it
+				log.Info("IPAddress already exists, another reconciler is working on it")
+				continue
+			}
+
+			ip, err := r.getIP(scope, pool)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
```

Comment:
I think you mentioned that you were concerned about the gap between this `getIP()` and the creation of the `IPAddress` containing it below, as a failure there would result in leaking the floating IP. Yes, it will. Unfortunately I can't currently think of anything which isn't more than a little Heath Robinson to do this robustly.

To do this robustly I think we would need to be able to identify after a failure that a floating IP which is not in the pool was intended to be in the pool. Currently we identify it because it is in an IPAddress owned by the pool. If it's not then we can't identify it.

The ideal solution would be to tag the floating IP, but unfortunately Neutron doesn't allow us to tag a floating IP on creation, only after creation. This means that we would still have a gap between floating IP creation and floating IP tagging where we could leak it. In practise I suspect that the creation of an IPAddress object below is likely much faster and more reliable than the tagging of a floating IP, so I suspect this would make reliability worse.

Other options are using the floating IP description to identify it as a member of the pool, or creating it in a multi-step process attached to a special port owned by the pool for the purpose:
* Create FIP attached to pool's 'special' port
* Tag port
* Unattach FIP from port

That way we can always determine the set of FIPs owned by the pool in OpenStack as the set of tagged FIPs plus any FIP attached to the pool's 'special' port. I believe this would be atomic, but it's extremely janky. I'm not sure I'd recommend it in the first instance.

We should probably tag all the FIPs anyway to aid cleanup if the control plane is lost for any reason.

How about we:
* Describe the potential for resource leakage in a big, comprehensive comment
* defer() the tagging of the FIP immediately after getIP so we prioritise the creation of the IPAddress, but we'll still try to tag the newly created FIP even if we exit with an error.
* Wrap the Create() call below in a client-go.util.retry.OnError(), or something better that takes a context
* update getIP() to look for an existing tagged FIP which isn't allocated to a member of the pool before creating a new one.